### ISSUE TITLE
Deleting many to many children from parent is not working

### ISSUE THREAD

**Issue by TanvirArjel** [dotnet/efcore#11055]
Here is the relative classes and configuration:

```C#
public class Product 
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long ProductId { get; set; }

        [Required, StringLength(50, MinimumLength = 3)]
        public string ProductName { get; set; }

        public ICollection<ProductSpecification> ProductSpecifications { get; set; }
    }


    public class ProductAttributeValue 
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long ProductAttributeValueId { get; set; }

        [Required, StringLength(20, MinimumLength = 3)]
        public string ProductAttributeValueName { get; set; }

        public ICollection<ProductSpecification> ProductSpecifications { get; set; }
    }
public class ProductSpecification
    {
        public long ProductId { get; set; }
        public long ProductAttributeValueId { get; set; }

        [StringLength(100)]
        public string Note { get; set; }

        public Product Product { get; set; }
        public ProductAttributeValue ProductAttributeValue { get; set; }
    }

//Configuration in the dbConext

 modelBuilder.Entity<ProductSpecification>().HasKey(ps => new { ps.ProductId, ps.ProductAttributeValueId });
            modelBuilder.Entity<ProductSpecification>().HasOne(p => p.Product).WithMany(ps => ps.ProductSpecifications).HasForeignKey(ps => ps.ProductId);
            modelBuilder.Entity<ProductSpecification>().HasOne(pav => pav.ProductAttributeValue).WithMany(ps => ps.ProductSpecifications).HasForeignKey(ps => ps.ProductAttributeValueId);


//Controller Method

public async Task<IActionResult> UpdateProduct([FromRoute] long id, [FromBody] Product product)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            if (id != product.ProductId)
            {
                return BadRequest();
            }

            Product productToBeUpdated = await _unitOfWork.Repository<Product>().GetEntityList(p => p.ProductId == id).Include(p => p.ProductSpecifications).SingleOrDefaultAsync();
            if (productToBeUpdated == null)
            {
                return NotFound();
            }
        
            foreach (ProductSpecification productSpecification in productToBeUpdated.ProductSpecifications.ToList())
            {
                productToBeUpdated.ProductSpecifications.Remove(productSpecification);
            }

            productToBeUpdated.ProductSpecifications = product.ProductSpecifications;

            productToBeUpdated.ModifiedOn = DateTime.UtcNow;
            await _unitOfWork.SaveChangesAsync();
            return Ok(true);
        }
```

I have also tried : 
```C#
foreach (ProductSpecification productSpecification in productToBeUpdated.ProductSpecifications.ToList())
            {
               _unitOfWork.Repository<ProductSpecifications>().DeleteEntity(productSpecification);
            }
```

Both of them throwing the following exception:
`The instance of entity type 'ProductSpecification' cannot be tracked because another instance with the same key value for {'ProductId', 'ProductAttributeValueId'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.`

**Comment by smitpatel** [comment]
` productToBeUpdated.ProductSpecifications = product.ProductSpecifications;`

Does the exception come from above line? Presently if you delete an entity, it is still in changetracker and you cannot add another entity with same identity. In 2.1.0-preview1 we changed it slightly to allow tracking another entity with same identity as deleted entity. I believe it should help in above case. See #7340 for more details on that functionality.

**Comment by TanvirArjel** [comment]
No, the exception not coming from that line..Exception coming from that line following lines of code:

```
foreach (ProductSpecification productSpecification in productToBeUpdated.ProductSpecifications.ToList())
            {
                productToBeUpdated.ProductSpecifications.Remove(productSpecification);
            } 
```

By the way, I finally I have found the exact reason of the exception and found the best way to solve the problem!

**Comment by ajcvickers** [comment]
@TanvirArjel Pretty sure that this is the same issue as @smitpatel referenced above, which is fixed in 2.1. However, it's really hard to see how this exception could come from that line of code. Can you please post a runnable project/solution or complete code listing that demonstrates this behavior so we can investigate if something else is going on?

**Comment by TanvirArjel** [comment]
In Entity Framework 6.x, during many to to many navigation properties update, we can clear/delete the existing child list and then add the new child list to the parent and finally update the parent and parent is updated with the new child list.

But in Entity Framework Core, we cannot do the same.The error message suggest that we cannot have same child with the same key value more than once. We have to maintain the uniqueness of entities in the Entity Framework transaction.

That's why we have to first identified which existing children are being deleted and which children are being newly added during update operation. This will make Entity framework Transaction state unique.

So the Update method should be as follows:

```
public async Task<IActionResult> UpdateProduct([FromRoute] long id, [FromBody] Product product)
{
    if (!ModelState.IsValid)
    {
       return BadRequest(ModelState);
    }

    if (id != product.ProductId)
    {
       return BadRequest();
    }

    Product productToBeUpdated = await _unitOfWork.Repository<Product>().GetEntityList(p => p.ProductId == id).Include(p => p.ProductSpecifications).SingleOrDefaultAsync();
    if (productToBeUpdated == null)
    {
        return NotFound();
    }

    _mapper.Map(product, productToBeUpdated); // If you use AutoMapper than ignore the many-to-many navigation property in the mapping

   List<ProductSpecification> productSpecificationsToBeDeleted = productToBeUpdated.ProductSpecifications.Except(product.ProductSpecifications, (c1, c2) => c1.ProductAttributeValueId == c2.ProductAttributeValueId).ToList();
   foreach (ProductSpecification productSpecification in productSpecificationsToBeDeleted)
   {
                productToBeUpdated.ProductSpecifications.Remove(productSpecification);
   }

   List<ProductSpecification> productSpecificationsToBeAdded = product.ProductSpecifications.Except(productToBeUpdated.ProductSpecifications, (c1, c2) => c1.ProductAttributeValueId == c2.ProductAttributeValueId).ToList();
   foreach (ProductSpecification productSpecification in productSpecificationsToBeAdded)
   {
                productToBeUpdated.ProductSpecifications.Add(productSpecification);
   }

   productToBeUpdated.ModifiedOn = DateTime.UtcNow;
   await _unitOfWork.SaveChangesAsync();
   return Ok(true);
}
```

Moreover we have to write an extension method for the linq Except() method as follows:
```
public static class LinqExcept
    {
        /// <summary>
        /// Returns all items in the first collection except the ones in the second collection that match the lambda condition
        /// </summary>
        /// <typeparam name="T">The type</typeparam>
        /// <param name="listA">The first list</param>
        /// <param name="listB">The second list</param>
        /// <param name="lambda">The filter expression</param>
        /// <returns>The filtered list</returns>
        public static IEnumerable<T> Except<T>(this IEnumerable<T> listA, IEnumerable<T> listB, Func<T, T, bool> lambda)
        {
            return listA.Except(listB, new LambdaComparer<T>(lambda));
        }

        /// <summary>
        /// Returns all items in the first collection that intersect the ones in the second collection that match the lambda condition
        /// </summary>
        /// <typeparam name="T">The type</typeparam>
        /// <param name="listA">The first list</param>
        /// <param name="listB">The second list</param>
        /// <param name="lambda">The filter expression</param>
        /// <returns>The filtered list</returns>
        public static IEnumerable<T> Intersect<T>(this IEnumerable<T> listA, IEnumerable<T> listB, Func<T, T, bool> lambda)
        {
            return listA.Intersect(listB, new LambdaComparer<T>(lambda));
        }
    }

```
